lib/connections/relay.go

```diff
@@ -0,0 +1,183 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package connections
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"sync"
+
+	"github.com/syncthing/syncthing/lib/osutil"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/relay/client"
+)
+
+var (
+	relayPriority = 200
+)
+
+func init() {
+	dialers["relay"] = relayDialer{}
+	listeners["relay"] = newRelayListener
+	listeners["dynamic+http"] = newRelayListener
+	listeners["dynamic+https"] = newRelayListener
+}
+
+type relayDialer struct{}
+
+func (relayDialer) Dial(id protocol.DeviceID, uri *url.URL, tlsCfg *tls.Config) (intermediateConnection, error) {
+	inv, err := client.GetInvitationFromRelay(uri, id, tlsCfg.Certificates)
+	if err != nil {
+		return intermediateConnection{}, err
+	}
+
+	conn, err := client.JoinSession(inv)
+	if err != nil {
+		return intermediateConnection{}, err
+	}
+
+	err = osutil.SetTCPOptions(conn.(*net.TCPConn))
+	if err != nil {
+		conn.Close()
+		return intermediateConnection{}, err
+	}
+
+	var tc *tls.Conn
+	if inv.ServerSocket {
+		tc = tls.Server(conn, tlsCfg)
+	} else {
+		tc = tls.Client(conn, tlsCfg)
+	}
+
+	err = tc.Handshake()
+	if err != nil {
+		tc.Close()
+		return intermediateConnection{}, err
+	}
+
+	return intermediateConnection{tc, "relay-dial", relayPriority}, nil
+}
+
+func (relayDialer) Priority() int {
+	return relayPriority
+}
+
+type relayListener struct {
+	uri     *url.URL
+	tlsCfg  *tls.Config
+	stopped chan struct{}
+	conns   chan intermediateConnection
+
+	err    error
+	client client.RelayClient
+	mut    sync.RWMutex
+}
+
+func (t *relayListener) Serve() {
+	t.mut.Lock()
+	t.err = nil
+	t.mut.Unlock()
+
+	clnt, err := client.NewClient(t.uri, t.tlsCfg.Certificates, nil)
+
+	go clnt.Serve()
+
+	if err != nil {
+		t.mut.Lock()
+		t.err = err
+		t.mut.Unlock()
+		l.Warnln("listen (BEP/relay):", err)
+		return
+	}
+
+	t.mut.Lock()
+	t.client = clnt
+	t.mut.Unlock()
+
+	for inv := range t.client.Invitations() {
+		conn, err := client.JoinSession(inv)
+		if err != nil {
+			l.Warnln("Joining relay session (BEP/relay):", err)
+			continue
+		}
+
+		err = osutil.SetTCPOptions(conn.(*net.TCPConn))
+		if err != nil {
+			l.Infoln(err)
+		}
+
+		var tc *tls.Conn
+		if inv.ServerSocket {
+			tc = tls.Server(conn, t.tlsCfg)
+		} else {
+			tc = tls.Client(conn, t.tlsCfg)
+		}
+
+		err = tc.Handshake()
+		if err != nil {
+			tc.Close()
+			l.Infoln("TLS handshake (BEP/relay):", err)
+			continue
+		}
+
+		t.conns <- intermediateConnection{tc, "relay-listen", relayPriority}
+	}
+
+	close(t.stopped)
+}
+
+func (t *relayListener) Stop() {
+	t.stopped = make(chan struct{})
+
+	t.mut.RLock()
+	clnt := t.client
+	t.mut.RUnlock()
+
+	clnt.Stop()
+	<-t.stopped
+}
+
+func (t *relayListener) WANAddresses() []*url.URL {
+	t.mut.RLock()
+	clnt := t.client
+	t.mut.RUnlock()
+
+	return []*url.URL{clnt.URI()}
+}
+
```

Comment:
The same implementation of `WANAddresses` and `LANAddresses` looks weird here. It mostly makes sense, though. One area that might be weird, is that I think relay addresses will get double added in `connections.AllAddresses`. I haven't chased the impact down all the way. It looks like it might double announce the relay connection through discovery. I don't think this is currently a bug, but is some extra work that might be weird.